### PR TITLE
FOUNDATIONS: Retrieve All Students

### LINKED ISSUE #30: FOUNDATIONS: Retrieve All Students

### PR DESCRIPTION
**Pull request by cacti79**
Closes  #30

### CODE REVIEW

**Comment by hassanhabib** [comment]
@ShriHumrudha can you please take a look at this PR?

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+using Xunit;
+
+namespace OtripleS.Portal.Web.Tests.Unit.Services.Students
+{
+    public partial class StudentServiceTests
+    {
+        [Theory]
+        [MemberData(nameof(CriticalApiExceptions))]
+        public async Task ShouldThrowCriticalDependencyExceptionOnRetrieveAllIfCriticalDependencyExceptionOccursAndLogItAsync(
+            Exception criticalDependencyException)
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs:23]: We shall stay consistent with the member data name - 'criticalApiException'

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            // given
+            var failedStudentDependencyException =
+                new FailedStudentDependencyException(
+                    criticalDependencyException);
+
+            var expectedStudentDepndencyException =
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs:30]: Typo

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedStudentDepndencyException))),
+                        Times.Once);
+
+            this.apiBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowStudentDependencyExceptionOnRetrieveAllIfDependencyApiErrorOccursAndLogItAsync()
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs:60]: As we are in the context of Student, we can ignore 'Student' here.

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            // given
+            var randomExceptionMessage = GetRandomString();
+            var responseMessage = new HttpResponseMessage();
+
+            var httpResponseException =
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs:66]: No need for parameter aliasing here.
Please verify in other places too.

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Logic.RetrieveAll.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+namespace OtripleS.Portal.Web.Tests.Unit.Services.Students
+{
+    public partial class StudentServiceTests
+    {
+        [Fact]
+        public async Task ShouldRetrieveAllStudentsAsync()
+        {
+            // given
+            List<Student> randomStudents = CreateRandomStudents();
+            List<Student> apiStudents = randomStudents;
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Logic.RetrieveAll.cs:23]: How about 'returningStudents' or 'retrievedStudents' ?

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Logic.RetrieveAll.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            List<Student> expetedStudents = apiStudents.DeepClone();
+
+            this.apiBrokerMock.Setup(broker =>
+                broker.GetAllStudentsAsync())
+                    .ReturnsAsync(apiStudents);
+
+            // when
+            List<Student> retrievedStudents =
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Logic.RetrieveAll.cs:31]: actualStudents?

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.cs [hunk 7]
```diff
@@ -29,10 +34,83 @@ public StudentServiceTests()
                 apiBroker: this.apiBrokerMock.Object,
                 loggingBroker: this.loggingBrokerMock.Object);
         }
+        public static TheoryData ValidationApiExceptions()
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.cs:37]: New line required.

File: OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedStudentDepndencyException))),
+                        Times.Once);
+
+            this.apiBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web.Tests.Unit/Services/Students/StudentServiceTests.Exceptions.RetrieveAll.cs:59]: Why is it not a theory? - DependencyApiExceptions

File: OtripleS.Portal.Web/Models/Students/Exceptions/FailedStudentDependencyException.cs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web/Models/Students/Exceptions/FailedStudentDependencyException.cs:1]: Header, here and other places too.

File: OtripleS.Portal.Web/Models/Students/Exceptions/FailedStudentDependencyException.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+using Xeptions;
+
+namespace OtripleS.Portal.Web.Models.Students.Exceptions
+{
+    public class FailedStudentDependencyException : Xeption
+    {
+        public FailedStudentDependencyException(Exception innerException)
+            : base("Failed student dependency error occurred, please contact support.", innerException)
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web/Models/Students/Exceptions/FailedStudentDependencyException.cs:9]: Parameter aliasing, here and other places too.

File: OtripleS.Portal.Web/Services/Students/IStudentService.cs [hunk 11]
```diff
@@ -11,5 +12,7 @@ namespace OtripleS.Portal.Web.Services.Students
     public interface IStudentService
     {
         ValueTask<Student> RegisterStudentAsync(Student student);
+
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web/Services/Students/IStudentService.cs:15]: New line space not needed

File: OtripleS.Portal.Web/Services/Students/StudentService.Exceptions.cs [hunk 12]
```diff
@@ -16,6 +17,8 @@ public partial class StudentService
     {
         private delegate ValueTask<Student> ReturningStudentFunction();
```
Comment (ShriHumrudha) [inline OtripleS.Portal.Web/Services/Students/StudentService.Exceptions.cs:19]: New line space not needed

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

**Comment by cacti79** [comment]
@ShriHumrudha @hassanhabib 
please take another look

**Pull request merged**